File: apps/web/src/lib/api/panel-security.ts
import { api, publicApi } from './client'
import { fetchNetworkAccess, type NetworkAccessStatus } from './network-access'
import type { PanelAccess } from './types'

export type PanelSecurityUrls = {
  entrance: string | null
  login: string | null
  local_login: string | null
  bare_host_note: string
}

export type ServerTimezoneInfo = {
  timezone: string
  label: string
  local_time: string
  utc_offset: string
}

export type TimezoneOption = {
  id: string
  label: string
}

export type PanelSecurityStatus = {
  security_entrance_enabled: boolean
  entry_path: string | null
  entry_path_display: string | null
  access_key_configured: boolean
  panel_host: string
  panel_port: number
  external_host: string | null
  panel_username: string | null
  urls: PanelSecurityUrls
  network: NetworkAccessStatus
  restart_required: boolean
  server_timezone: ServerTimezoneInfo
  timezone_options: TimezoneOption[]
}

export type PanelSecurityUpdateBody = {
  external_host?: string | null
  entry_path?: string | null
  regenerate_entry?: boolean
  regenerate_access_key?: boolean
  enable_entrance?: boolean
  username?: string
  password?: string
  timezone?: string
}

export type PanelSecurityUpdateResponse = {
  ok: boolean
  messages: string[]
  access_key: string | null
  status: PanelSecurityStatus
  restart_required: boolean
}

function isNotFoundError(err: unknown): boolean {
  return err instanceof Error && err.message.includes('404')
}

async function fetchPanelSecurityLegacy(): Promise<PanelSecurityStatus> {
  const [network, access] = await Promise.all([
    fetchNetworkAccess(),
    publicApi<PanelAccess>('/panel/access'),
  ])
  const entry = access.entry_path?.trim() || null
  const host = network.external_host ?? access.access_ip
  const port = network.port

  return {
    security_entrance_enabled: Boolean(entry),
    entry_path: entry,
    entry_path_display: entry ? `/${entry}` : null,
    access_key_configured: false,
    panel_host: network.bind_host,
    panel_port: port,
    external_host: network.external_host,
    panel_username: null,
    urls: {
      entrance: access.entrance_url ?? null,
      login: access.panel_url ?? null,
      local_login: null,
      bare_host_note: entry
        ? `http://${host}:${port} returns 404 when security entrance is enabled`
        : `http://${host}:${port}/ui/ — standard panel URL`,
    },
    network,
    restart_required: false,
    server_timezone: {
      timezone: 'UTC',
      label: 'UTC',
      local_time: '',
      utc_offset: '+00:00',
    },
    timezone_options: [{ id: 'UTC', label: 'UTC' }],
  }
}

export async function fetchPanelSecurity(): Promise<PanelSecurityStatus> {
  try {
    return await api<PanelSecurityStatus>('/deploy/security')
  } catch (err) {
    if (!isNotFoundError(err)) throw err
    return fetchPanelSecurityLegacy()
  }
}

export function updatePanelSecurity(body: PanelSecurityUpdateBody) {
  return api<PanelSecurityUpdateResponse>('/deploy/security', {
    method: 'PATCH',
    body: JSON.stringify(body),
  })
}

File: src/server/schemas/panel_security.py
"""Panel security / network configuration schemas."""

from __future__ import annotations

from pydantic import BaseModel, Field

from server.schemas.network import NetworkAccessStatusResponse


class PanelSecurityUrls(BaseModel):
    entrance: str | None = None
    entrance_access: str | None = None
    login: str | None = None
    local_login: str | None = None
    bare_host_note: str = "http://<host>:<port> returns 404 when security entrance is enabled"


class ServerTimezoneInfo(BaseModel):
    timezone: str = "UTC"
    label: str = "UTC"
    local_time: str = ""
    utc_offset: str = "+00:00"


class TimezoneOption(BaseModel):
    id: str
    label: str


class PanelSecurityStatusResponse(BaseModel):
    security_entrance_enabled: bool
    entry_path: str | None = None
    entry_path_display: str | None = None
    access_key_configured: bool = False
    panel_host: str
    panel_port: int
    external_host: str | None = None
    panel_username: str | None = None
    urls: PanelSecurityUrls
    network: NetworkAccessStatusResponse
    restart_required: bool = False
    server_timezone: ServerTimezoneInfo
    timezone_options: list[TimezoneOption] = Field(default_factory=list)


class PanelSecurityUpdateRequest(BaseModel):
    external_host: str | None = Field(
        default=None,
        description="Public domain or IP for access URLs (empty clears)",
    )
    entry_path: str | None = Field(
        default=None,
        description="8-char hex entrance path, or off/false/disabled to disable",
    )
    regenerate_entry: bool = False
    regenerate_access_key: bool = False
    enable_entrance: bool | None = None
    username: str | None = Field(default=None, min_length=1, max_length=64)
    password: str | None = Field(default=None, min_length=8, max_length=128)
    timezone: str | None = Field(
        default=None,
        description="IANA timezone for cron schedules and server time display",
    )


class PanelSecurityUpdateResponse(BaseModel):
    ok: bool = True
    messages: list[str] = Field(default_factory=list)
    access_key: str | None = Field(
        default=None,
        description="Plain access key when regenerated (show once)",
    )
    status: PanelSecurityStatusResponse
    restart_required: bool = False
